frankenstein/chunk_store.go

```diff
@@ -0,0 +1,379 @@
+	}
+	password, _ := url.User.Password()
+	creds := credentials.NewStaticCredentials(url.User.Username(), password, "")
+	config := aws.NewConfig().WithCredentials(creds)
+	if strings.Contains(url.Host, ".") {
+		config = config.WithEndpoint(fmt.Sprintf("http://%s", url.Host)).WithRegion("dummy")
+	} else {
+		config = config.WithRegion(url.Host)
+	}
+	return config, nil
+}
+
+type chunkStore struct {
+	dynamodb   *dynamodb.DynamoDB
+	s3         *s3.S3
+	tableName  string
+	bucketName string
+	cfg        ChunkStoreConfig
+}
+
+func hours(from, through model.Time) []model.Time {
+	// TODO it
+	return nil
+}
+
+func (c *chunkStore) Put(chunks []Chunk) error {
+	// TODO: parallelise
+	for _, chunk := range chunks {
+		err := timeRequest("Put", s3RequestDuration, func() error {
+			var err error
+			_, err = c.s3.PutObject(&s3.PutObjectInput{
+				Body:   bytes.NewReader(chunk.Data),
+				Bucket: aws.String(c.bucketName),
+				Key:    aws.String(chunk.ID),
+			})
+			return err
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	writeReqs := []*dynamodb.WriteRequest{}
+	for _, chunk := range chunks {
+		for _, hour := range hours(chunk.From, chunk.Through) {
+			for label, value := range chunk.Metric {
+				// TODO escaping
+				// TODO compression
+				hashValue := fmt.Sprintf("%s:%s", hour, label)
+				rangeValue := fmt.Sprintf("%s=%s:%s", label, value, chunk.ID)
+				chunkValue, err := json.Marshal(chunk)
+				if err != nil {
+					return err
+				}
+
+				writeReqs = append(writeReqs, &dynamodb.WriteRequest{
+					PutRequest: &dynamodb.PutRequest{
+						Item: map[string]*dynamodb.AttributeValue{
+							hashKey:  {S: aws.String(hashValue)},
+							rangeKey: {S: aws.String(rangeValue)},
+							chunkKey: {B: chunkValue},
+						},
+					},
+				})
+			}
+		}
+	}
+
+	var resp *dynamodb.BatchWriteItemOutput
+	err := timeRequest("BatchWriteItem", dynamoRequestDuration, func() error {
+		var err error
+		resp, err = c.dynamodb.BatchWriteItem(&dynamodb.BatchWriteItemInput{
+			RequestItems:           map[string][]*dynamodb.WriteRequest{c.tableName: writeReqs},
+			ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+		})
+		return err
+	})
+	for _, cc := range resp.ConsumedCapacity {
+		dynamoConsumedCapacity.WithLabelValues("PutItem").
+			Add(float64(*cc.CapacityUnits))
+	}
+	return err
+}
+
+func (c *chunkStore) Get(from, through model.Time, matchers ...*metric.LabelMatcher) ([]Chunk, error) {
+	chunks, err := c.consultChunkIndex(from, through, matchers)
+	if err != nil {
+		return nil, err
+	}
+	return c.fetchChunkData(chunks)
+}
+
+func (c *chunkStore) consultChunkIndex(from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
+	incomingChunkSets := make(chan []Chunk)
+	incomingErrors := make(chan error)
+	hs := hours(from, through)
+	for _, hour := range hs {
+		for _, matcher := range matchers {
+			go func(hour model.Time, matcher *metric.LabelMatcher) {
```

Comment:
Could you extract this into a separate function? The indentation level is getting a bit unwieldy below.